juju/juju
# BridgePolicy - Deal With SpaceInfo Instead of Space Names

## Description of change

This patch changes `BridgePolicy` so that we begin to pass around `SpaceInfo` instead of space name strings. It is not the final form of this logic - completion depends upon finalisation of subnet space designation being by ID instead of name.

A space cache is created upon `BridgePolicy` instantiation, which is used in this intermediate form to look up spaces by space names.

## QA steps

Bridging via Constraints:

- Bootstrap guimaas.
- `juju add-machine --constraints spaces=default,space-alt2`.
- `juju add-machine lxd:0 --constraints spaces=default,space-alt2`.
- Once the container has started, `juju ssh` to the host machine and check that bridges have been created for the NIC in each space.

Bridging via Endpoints:

- Bootstrap guimaas (or destroy previous model and create a new one).
- `juju add-machine --constraints spaces=default,space-alt2`.
- `juju deploy keystone --bind "public=default internal=space-alt2 admin=default shared-db=space-alt2" --to lxd:0`
- Once the container has started, `juju ssh` to the host machine and check that bridges have been created for the NIC in each space.

## Documentation changes

None.

## Bug reference

N/A
